cl/corpus_importer/signals.py

```diff
@@ -0,0 +1,122 @@
+from functools import partial
+
+from django.conf import settings
+from django.db import transaction
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+
+from cl.corpus_importer.tasks import make_docket_by_iquery
+from cl.lib.redis_utils import (
+    acquire_atomic_redis_lock,
+    get_redis_interface,
+    make_update_pacer_case_id_key,
+    release_atomic_redis_lock,
+)
+from cl.search.models import Court, Docket
+
+
+def update_latest_case_id_and_schedule_iquery_sweep(docket: Docket) -> None:
+    """Updates the latest PACER case ID and schedules iquery retrieval tasks.
+
+    :param docket: The incoming Docket instance.
+    :return: None
+    """
+
+    r = get_redis_interface("CACHE")
+    court_id = docket.court.pk
+    # Get the latest pacer_case_id from Redis using a lock to avoid race conditions
+    # when getting and updating it.
+    update_lock_key = make_update_pacer_case_id_key(court_id)
+    # ttl one hour.
+    lock_value = acquire_atomic_redis_lock(r, update_lock_key, 60 * 60 * 1000)
+
+    current_iquery_pacer_case_id_final = int(
+        r.hget("iquery_pacer_case_id_final", court_id) or 0
+    )
+    iquery_pacer_case_id_status = int(
+        r.hget("iquery_pacer_case_id_status", court_id) or 0
+    )
+    incoming_pacer_case_id = int(docket.pacer_case_id)
+    updated_pacer_case_id = False
+    if incoming_pacer_case_id > current_iquery_pacer_case_id_final:
+        r.hset("iquery_pacer_case_id_final", court_id, incoming_pacer_case_id)
+        updated_pacer_case_id = True
+
+    if updated_pacer_case_id:
+        task_scheduled_countdown = 0
+
+        while iquery_pacer_case_id_status + 1 < incoming_pacer_case_id:
+            tasks_processed_in_this_batch = 0
+
+            # Schedule tasks in batches of IQUERY_SWEEP_BATCH_SIZE to avoid
+            # a celery runaway scheduling tasks with countdowns larger than
+            # the celery visibility_timeout.
+            while (
+                tasks_processed_in_this_batch
+                < settings.IQUERY_SWEEP_BATCH_SIZE
+                and iquery_pacer_case_id_status + 1 < incoming_pacer_case_id
+            ):
```

Comment:
Let's remove the batch size protection and do like you say to prevent new dockets from triggering the signal. I think that's easier than having this complexity with us forever, right?

Reply:
Yeah, this is better now as it will slowly catch up with the courts. I removed and also added the setting `IQUERY_SWEEP_UPLOADS_SIGNAL_ENABLED`, set to `False` by default. Once we catch up with the courts, we can set it to `True`.